mono/metadata/threads.c

```diff
@@ -4996,6 +4998,68 @@ mono_thread_is_foreign (MonoThread *thread)
 	return info->runtime_thread == FALSE;
 }
 
+#ifndef HOST_WIN32
+static void
+threads_native_thread_join_lock (gpointer key, gpointer value)
+{
+	pthread_t thread = (pthread_t)key;
+	if (thread != pthread_self ()) {
+		MONO_ENTER_GC_SAFE;
+		/* This shouldn't block */
+		mono_threads_join_lock ();
+		mono_native_thread_join (thread);
+		mono_threads_join_unlock ();
+		MONO_EXIT_GC_SAFE;
+	}
+}
+inline static void
+threads_native_thread_join_nolock (gpointer key, gpointer value)
+{
+	pthread_t thread = (pthread_t)key;
+	MONO_ENTER_GC_SAFE;
+	mono_native_thread_join (thread);
+	MONO_EXIT_GC_SAFE;
+}
+
+inline static void
+threads_add_joinable_thread_nolock (gpointer key)
+{
+	g_hash_table_insert (joinable_threads, key, key);
+}
+#else
+static void
+threads_native_thread_join_lock (gpointer key, gpointer value)
+{
```

Comment:
Why can't we use `mono_native_thread_join` on both platforms?

Reply:
wanted to avoid additional additional openthread since we already have a valid thread handle in the table. mono_native_thread_join only accepts tid.

Reply:
Moved to a handle base join in the windows threading backend in complement to the already existing tid version.